javascript.md

```diff
@@ -0,0 +1,793 @@
+
+Module systems also provide us with dependency injection patterns, which are crucial when it comes to testing individual components in isolation.
+
+## Strict Mode
+
+**Always** put [`'use strict';`][4] at the top of your modules. Strict mode allows you to catch nonsensical behavior, discourages poor practices, and _is faster_ because it allows compilers to make certain assumptions about your code.
+
+## Spacing
+
+Spacing must be consistent across every file in the application. To this end, using something like [`.editorconfig`][5] configuration files is highly encouraged. Here are the defaults we suggest to get started with JavaScript indentation.
+
+```ini
+# editorconfig.org
+root = true
+
+[*]
+indent_style = space
+indent_size = 2
+end_of_line = lf
+charset = utf-8
+trim_trailing_whitespace = true
+insert_final_newline = true
+
+[*.md]
+trim_trailing_whitespace = false
+```
+
+We recommend using 2 spaces for indentation. The `.editorconfig` file can take care of that for us and everyone would be able to create the correct tabs by pressing the tab key.
+
+Spacing doesn't just entail tabbing, but also the spaces before, after, and in between arguments of a function declaration. This kind of spacing is typically highly irrelevant to get right, and it'll be hard for most teams to even arrive at a scheme that will satisfy everyone.
+
+```js
+function () {}
+```
+
+```js
+function( a, b ){}
+```
+
+```js
+function(a, b) {}
+```
+
+```js
+function (a,b) {}
+```
+
+Try to keep these differences to a minimum, but don't put much thought to it either.
+
+Where possible, improve readability by keeping lines below the 80-character mark.
+
+## Naming
+
+- Functions, variables, methods, objects and instances should be named using `camelCase`.
+- Constructors and prototypes should use `UpperCamelCase`.
+- Symbolic constants should use `UPPERCASE`.
+
+## Semicolons`;`
+
+The majority of JavaScript developers [prefer using semicolons][6] at the end of lines and we recommend following this practice. This choice is done to avoid potential issues with Automatic Semicolon Insertion _(ASI)_.
+
+## Strings
+
+Strings should always be quoted using the same quotation mark. Use `'` or `"` consistently throughout your codebase. Ensure the team is using the same quotation mark in every portion of JavaScript that's authored. In general, we prefer to use sing quotation parks `'`.
+
+##### Bad
+
+```js
+var message = 'oh hai ' + name + "!";
+```
+
+##### Good
+
+```js
+var message = 'oh hai ' + name + '!';
+```
+
+Usually you'll be a happier JavaScript developer if you hack together a parameter-replacing method like [`util.format` in Node][12]. That way it'll be far easier to format your strings, and the code looks a lot cleaner too.
+
+##### Better
```

Comment:
To keep this lean, I'm wondering if this `Better` section is getting out of scope. The idea of the `Strings` section seems to be about not mixing quotation styles, but showing optimized implementations seems better left to another document. Perhaps provide a separate "cookbook" style document with code snippets that this links to?